comp/api/api/apiimpl/observability/telemetry.go

```diff
@@ -22,30 +22,46 @@ const (
 	metricHelp      = "Request duration distribution by server, method, path, and status (in seconds)."
 )
 
-// TelemetryHandler is a middleware which sends a metric representing the duration of the request,
-// tagged with the given server name, response status code, path, and method.
-func TelemetryHandler(telemetry telemetry.Component, serverName string) mux.MiddlewareFunc {
-	return telemetryHandler(telemetry, serverName, clock.New())
+type telemetryMiddlewareFactory struct {
+	requestDuration telemetry.Histogram
+	clock           clock.Clock
 }
 
-func telemetryHandler(telemetry telemetry.Component, serverName string, clock clock.Clock) mux.MiddlewareFunc {
-	tags := []string{"servername", "status_code", "method", "path"}
-	var buckets []float64 // use default buckets
-	requestDuration := telemetry.NewHistogram(metricSubsystem, metricName, tags, metricHelp, buckets)
+// TelemetryMiddlewareFactory creates a telemetry middleware tagged with a given server name
+type TelemetryMiddlewareFactory interface {
+	Middleware(serverName string) mux.MiddlewareFunc
+}
 
+func (th *telemetryMiddlewareFactory) Middleware(serverName string) mux.MiddlewareFunc {
 	return func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 			var code int
 			next = extractStatusHandler(&code)(next)
 
 			var duration time.Duration
-			next = timeHandler(clock, &duration)(next)
+			next = timeHandler(th.clock, &duration)(next)
 
 			next.ServeHTTP(w, r)
 
 			path := extractPath(r)
-			metricWithTags := requestDuration.WithValues(serverName, strconv.Itoa(code), r.Method, path)
+			metricWithTags := th.requestDuration.WithValues(serverName, strconv.Itoa(code), r.Method, path)
 			metricWithTags.Observe(duration.Seconds())
```

Comment:
It seems we can simplify the call using `requestDuration.Observer(duration.Seconds(),  serverName, strconv.Itoa(code), r.Method, path)`

https://github.com/DataDog/datadog-agent/blob/main/comp/core/telemetry/telemetryimpl/prom_histogram.go#L19-L21

That way we avoid creating an extra object per metric invokation. 

What do you think?

Reply:
Oh yep I didn't see that one, nice 👍